File: app/src/main/java/com/example/vladimir/easyenglishlearn/category_edit/CategoryEditViewModel.java
package com.example.vladimir.easyenglishlearn.category_edit;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.databinding.ObservableField;
import android.support.annotation.StringRes;
import android.text.TextUtils;

import com.example.vladimir.easyenglishlearn.R;
import com.example.vladimir.easyenglishlearn.SingleLiveEvent;
import com.example.vladimir.easyenglishlearn.db.CategoryRepository;
import com.example.vladimir.easyenglishlearn.db.CategoryRepositoryImpl;
import com.example.vladimir.easyenglishlearn.model.Word;

import java.util.List;

import static com.example.vladimir.easyenglishlearn.Constants.EMPTY_STRING;

public class CategoryEditViewModel extends ViewModel {

    public final ObservableField<String> categoryName = new ObservableField<>();
    public final ObservableField<String> lexeme = new ObservableField<>();
    public final ObservableField<String> translation = new ObservableField<>();
    private SingleLiveEvent<Integer> mMessageLiveData;
    private SingleLiveEvent<Void> mFragmentCloseLiveData;
    private MutableLiveData<List<Word>> mWordsLiveData;
    private CategoryRepository mRepository;
    private String mOldCategoryName;
    private int mWordIndex;


    public CategoryEditViewModel(String categoryName) {
        mRepository = CategoryRepositoryImpl.getInstance();
        mOldCategoryName = categoryName;
        this.categoryName.set(categoryName);
        mWordsLiveData = new MutableLiveData<>();
        mWordsLiveData.setValue(mRepository.getWordsByCategory(categoryName));
        mMessageLiveData = new SingleLiveEvent<>();
        mFragmentCloseLiveData = new SingleLiveEvent<>();
        cleanTextFields();
    }

    @SuppressWarnings("ConstantConditions")
    public void onBtnSaveCategoryClick() {
        String newCategoryName = categoryName.get().trim();
        if (TextUtils.isEmpty(newCategoryName)) {
            showMessage(R.string.cef_toast_save_edit_category);
        } else {
            if (TextUtils.isEmpty(mOldCategoryName)) {
                addNewCategory(newCategoryName, mWordsLiveData.getValue());
            } else {
                updateCategory(mOldCategoryName, newCategoryName, mWordsLiveData.getValue());
            }
            mFragmentCloseLiveData.call();
        }
    }

    @SuppressWarnings("ConstantConditions")
    public void onBtnSaveWordClick() {
        if (isTextFieldsNotEmpty()) {
            Word newWord = new Word(lexeme.get().trim(), translation.get().trim());
            List<Word> wordList = mWordsLiveData.getValue();
            if (mWordIndex >= 0) {
                wordList.set(mWordIndex, newWord);
            } else {
                wordList.add(newWord);
            }
            mWordsLiveData.setValue(wordList);
            cleanTextFields();
        } else {
            showMessage(R.string.cef_toast_save_word_empty_fields);
        }
    }

    public void onBtnCleanClick() {
        cleanTextFields();
    }

    @SuppressWarnings("ConstantConditions")
    public void onIconRemoveWordClick(Word word) {
        List<Word> wordList = mWordsLiveData.getValue();
        wordList.remove(word);
        mWordsLiveData.setValue(wordList);
        cleanTextFields();
    }

    @SuppressWarnings("ConstantConditions")
    public void onRvItemClick(Word word) {
        lexeme.set(word.getLexeme());
        translation.set(word.getTranslation());
        mWordIndex = mWordsLiveData.getValue().indexOf(word);
    }

    @SuppressWarnings("ConstantConditions")
    private boolean isTextFieldsNotEmpty() {
        return !TextUtils.isEmpty(categoryName.get().trim()) &&
                !TextUtils.isEmpty(lexeme.get().trim()) &&
                !TextUtils.isEmpty(translation.get().trim());
    }

    private void cleanTextFields() {
        lexeme.set(EMPTY_STRING);
        translation.set(EMPTY_STRING);
        mWordIndex = -1;
    }

    private void showMessage(@StringRes int resId) {
        mMessageLiveData.setValue(resId);
    }

    private void addNewCategory(String categoryName, List<Word> wordList) {
        mRepository.addNewCategory(categoryName, wordList);
    }

    private void updateCategory(String oldCategoryName, String newCategoryName, List<Word> wordList) {
        mRepository.updateCategory(oldCategoryName, newCategoryName, wordList);
    }

    LiveData<List<Word>> getWordsLiveData() {
        return mWordsLiveData;
    }

    LiveData<Integer> getMessageLiveData() {
        return mMessageLiveData;
    }

    LiveData<Void> getFragmentCloseLiveData() {
        return mFragmentCloseLiveData;
    }
}

File: app/src/main/java/com/vladimir_x/easyenglishlearn/ui/category_edit/CategoryEditViewModel.kt
package com.vladimir_x.easyenglishlearn.ui.category_edit

import android.text.TextUtils
import androidx.annotation.StringRes
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vladimir_x.easyenglishlearn.Constants
import com.vladimir_x.easyenglishlearn.R
import com.vladimir_x.easyenglishlearn.domain.WordsInteractor
import com.vladimir_x.easyenglishlearn.model.Word
import com.vladimir_x.easyenglishlearn.util.SingleLiveEvent
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

@HiltViewModel
class CategoryEditViewModel @Inject constructor(
    state: SavedStateHandle,
    private val wordsInteractor: WordsInteractor
) : ViewModel() {
    private val _messageLiveData: SingleLiveEvent<Int> = SingleLiveEvent()
    private val _fragmentCloseLiveData: SingleLiveEvent<Unit> = SingleLiveEvent()
    private val _wordsLiveData: MutableLiveData<List<Word>> = MutableLiveData<List<Word>>()
    private val _currentWordLiveData: MutableLiveData<Pair<String, String>> =
        MutableLiveData<Pair<String, String>>()
    private var oldCategoryName: String = ""
    private var wordIndex = 0
    private var categoryName: String = ""
    private var lexeme: String = ""
    private var translation: String = ""

    val wordsLiveData: LiveData<List<Word>>
        get() = _wordsLiveData

    val currentWordLiveData: LiveData<Pair<String, String>>
        get() = _currentWordLiveData

    val messageLiveData: LiveData<Int?>
        get() = _messageLiveData

    val fragmentCloseLiveData: LiveData<Unit?>
        get() = _fragmentCloseLiveData

    init {
        val categoryName = state.get<String>(Constants.ARG_CATEGORY_NAME)
        categoryName?.let {
            this.categoryName = categoryName
            this.oldCategoryName = categoryName
            cleanTextFields()
            subscribeWordsToData()
        }
    }

    fun onBtnSaveCategoryClick(categoryName: String) {
        val newCategoryName: String = categoryName.trim()
        if (TextUtils.isEmpty(newCategoryName)) {
            showMessage(R.string.cef_toast_save_edit_category)
        } else {
            if (TextUtils.isEmpty(oldCategoryName)) {
                _wordsLiveData.value?.let { addNewCategory(newCategoryName, it) }
            } else {
                _wordsLiveData.value?.let { updateCategory(oldCategoryName, newCategoryName, it) }
            }
            _fragmentCloseLiveData.call()
        }
    }

    fun onBtnSaveWordClick(categoryName: String, lexeme: String, translation: String) {
        this.categoryName = categoryName
        this.lexeme = lexeme
        this.translation = translation
        if (isTextFieldsNotEmpty) {
            val newWord = Word(lexeme.trim(), translation.trim())
            _wordsLiveData.value?.let { list ->
                val newList = mutableListOf<Word>().apply {
                    addAll(list)
                }
                if (wordIndex >= 0) {
                    newList[wordIndex] = newWord
                } else {
                    newList.add(newWord)
                }
                _wordsLiveData.value = newList
                cleanTextFields()
            }
        } else {
            showMessage(R.string.cef_toast_save_word_empty_fields)
        }
    }

    fun onBtnCleanClick() {
        cleanTextFields()
    }

    fun onIconRemoveWordClick(word: Word) {
        _wordsLiveData.value?.let { list ->
            val newList = mutableListOf<Word>().apply {
                addAll(list)
            }
            newList.remove(word)
            _wordsLiveData.value = newList
            cleanTextFields()
        }
    }

    fun onItemClick(word: Word) {
        lexeme = word.lexeme
        translation = word.translation
        wordIndex = _wordsLiveData.value?.indexOf(word) ?: -1
        updateCurrentWord()
    }

    private val isTextFieldsNotEmpty: Boolean
        get() = !TextUtils.isEmpty(categoryName.trim()) &&
                !TextUtils.isEmpty(lexeme.trim()) &&
                !TextUtils.isEmpty(translation.trim())

    private fun subscribeWordsToData() {
        viewModelScope.launch {
            _wordsLiveData.value = wordsInteractor.getWordsByCategory(oldCategoryName)
        }
    }

    private fun cleanTextFields() {
        lexeme = Constants.EMPTY_STRING
        translation = Constants.EMPTY_STRING
        wordIndex = -1
        updateCurrentWord()
    }

    private fun showMessage(@StringRes resId: Int) {
        _messageLiveData.value = resId
    }

    private fun addNewCategory(categoryName: String, wordList: List<Word>) {
        viewModelScope.launch {
            withContext(Dispatchers.IO) {
                wordsInteractor.addNewCategory(wordList, categoryName)
            }
            showMessage(R.string.category_added)
        }
    }

    private fun updateCategory(
        oldCategoryName: String,
        newCategoryName: String,
        wordList: List<Word>
    ) {
        viewModelScope.launch {
            withContext(Dispatchers.IO) {
                wordsInteractor.updateCategory(
                    oldCategoryName,
                    newCategoryName,
                    wordList
                )
            }
            showMessage(R.string.category_edited)
        }
    }

    private fun updateCurrentWord() {
        _currentWordLiveData.value = Pair(lexeme, translation)
    }
}